File: packages/frontend/src/parser/AST.ts
import { Token } from '@nox/shared';

export type Stmt = 
    | LetStmt 
    | AssignStmt 
    | PrintStmt 
    | WhileStmt 
    | ForStmt
    | IfStmt 
    | MatchStmt
    | FnStmt 
    | ReturnStmt 
    | ExprStmt;

export interface LetStmt {
    type: 'LetStmt';
    name: string;
    initializer: Expr;
    isMutable: boolean;
}

export interface AssignStmt {
    type: 'AssignStmt';
    name: string;
    value: Expr;
}

export interface PrintStmt {
    type: 'PrintStmt';
    expression: Expr;
}

export interface WhileStmt {
    type: 'WhileStmt';
    condition: Expr;
    body: Stmt[];
}

export interface ForStmt {
    type: 'ForStmt';
    item: string;
    iterable: Expr;
    body: Stmt[];
}

export interface MatchStmt {
    type: 'MatchStmt';
    expression: Expr;
    cases: { condition: Expr | null, body: Stmt[] }[];
}

export interface IfStmt {
    type: 'IfStmt';
    condition: Expr;
    thenBranch: Stmt[];
    elseBranch?: Stmt[];
}

export interface FnStmt {
    type: 'FnStmt';
    name: string;
    params: string[];
    body: Stmt[];
}

export interface ReturnStmt {
    type: 'ReturnStmt';
    value: Expr;
}

export interface ExprStmt {
    type: 'ExprStmt';
    expression: Expr;
}

export type Expr = 
    | BinaryExpr 
    | UnaryExpr 
    | LiteralExpr 
    | VariableExpr 
    | CallExpr 
    | GroupingExpr
    | ArgCountExpr
    | DictExpr
    | RangeExpr;

export interface DictExpr {
    type: 'DictExpr';
    entries: { key: string, value: Expr }[];
}

export interface RangeExpr {
    type: 'RangeExpr';
    start: Expr;
    end: Expr;
}

export interface BinaryExpr {
    type: 'BinaryExpr';
    left: Expr;
    operator: Token;
    right: Expr;
}

export interface UnaryExpr {
    type: 'UnaryExpr';
    operator: Token;
    right: Expr;
    kind: 'prefix' | 'postfix';
}

export interface LiteralExpr {
    type: 'LiteralExpr';
    value: number | string | boolean | null;
}

export interface VariableExpr {
    type: 'VariableExpr';
    name: string;
}

export interface CallExpr {
    type: 'CallExpr';
    callee: string;
    args: Expr[];
}

export interface GroupingExpr {
    type: 'GroupingExpr';
    expression: Expr;
}

export interface ArgCountExpr {
    type: 'ArgCountExpr';
}

File: packages/frontend/src/ast.rs
use nox_shared::Token;

#[derive(Debug, Clone)]
pub enum Stmt {
    Let { name: String, initializer: Expr, is_mutable: bool },
    Assign { name: String, value: Expr },
    Print { expression: Expr },
    While { condition: Expr, body: Vec<Stmt> },
    For { item: String, iterable: Expr, body: Vec<Stmt> },
    If { condition: Expr, then_branch: Vec<Stmt>, else_branch: Option<Vec<Stmt>> },
    Match { expression: Expr, cases: Vec<(Option<Expr>, Vec<Stmt>)> },
    Fn { name: String, params: Vec<String>, body: Vec<Stmt> },
    Return { value: Expr },
    Expr(Expr),
}

#[derive(Debug, Clone)]
pub enum Expr {
    Binary { left: Box<Expr>, operator: Token, right: Box<Expr> },
    Unary { operator: Token, right: Box<Expr>, is_prefix: bool },
    Literal(LiteralValue),
    Variable(String),
    Call { callee: String, args: Vec<Expr> },
    Grouping(Box<Expr>),
    ArgCount,
    Dict(Vec<(String, Expr)>),
    Range { start: Box<Expr>, end: Box<Expr> },
    Pipe { left: Box<Expr>, right: Box<Expr> },
    SafeCall { left: Box<Expr>, right: String },
    NullCoalesce { left: Box<Expr>, right: Box<Expr> },
}

#[derive(Debug, Clone, PartialEq)]
pub enum LiteralValue {
    Number(i64),
    String(String),
    Boolean(bool),
    Null,
}
